### ISSUE TITLE
Surface Book 2 15": No DGPU support?

### ISSUE THREAD

**Issue by polarizing** [jakeday/linux-surface#57]
Everything is working nicely except there is no support for dGPU? I installed the NVIDIA binary driver 387 from software & updates and it works without using your kernel. After installing your kernel, nvidia-smi shows command not found?

**Comment by creatornader** [comment]
I am having the exact same issue. Drivers and GPU worked perfectly before installing kernel. Now, after kernel, nvidia-smi shows command not found.

**Comment by jakeday** [comment]
@polarizing @creatornader The command not found error shouldn't be tied to whether or not your dgpu is being detected and a module loaded. That's from the nvidia driver itself. If you're using secure boot, you're likely to see that error. I'm confused by the fact that you said it was working prior to installing my kernel. Was there any other changes? Nothing in the kernel would impact that. You can see if your device is detected and module loaded by running this command:

$ sudo lshw -C display

**Comment by jakeday** [comment]
My theory on this is that the display will not switch to the nvidia module as the i915 is required for the display since the ipts modules are loaded. While the nvidia module does get loaded, it's not associated with a display (as seen with the command I posted previously), and the nvidia tools won't properly detect it. I would imagine disabling the i915 and ipts modules, and associating the display with your nvidia driver will make it work.

**Comment by creatornader** [comment]
I think you may be right. Here is the output when I run the command you mentioned "sudo lshw -C display":
  *-display               
       description: VGA compatible controller
       product: Intel Corporation
       vendor: Intel Corporation
       physical id: 2
       bus info: pci@0000:00:02.0
       version: 07
       width: 64 bits
       clock: 33MHz
       capabilities: pciexpress msi pm vga_controller bus_master cap_list rom
       configuration: driver=i915 latency=0
       resources: irq:135 memory:d3000000-d3ffffff memory:b0000000-bfffffff ioport:4000(size=64) memory:c0000-dffff
  *-display
       description: 3D controller
       product: NVIDIA Corporation
       vendor: NVIDIA Corporation
       physical id: 0
       bus info: pci@0000:02:00.0
       version: a1
       width: 64 bits
       clock: 33MHz
       capabilities: pm msi pciexpress bus_master cap_list rom
       configuration: driver=nouveau latency=0
       resources: irq:127 memory:d2000000-d2ffffff memory:c0000000-cfffffff memory:d0000000-d1ffffff ioport:3000(size=128) memory:90000000-9007ffff

**Comment by creatornader** [comment]
And here's what I get when I run "dpkg -l | grep nvidia":

rc  nvidia-381                                  381.22-0ubuntu0~gpu16.04.2                   amd64        NVIDIA binary driver - version 381.22
ii  nvidia-387                                  387.34-0ubuntu0~gpu16.04.2                   amd64        NVIDIA binary driver - version 387.34
ii  nvidia-docker                               1.0.1-1                                      amd64        NVIDIA Docker container tools
ii  nvidia-modprobe                             361.28-1                                     amd64        utility to load NVIDIA kernel modules and create device nodes
rc  nvidia-opencl-icd-381                       381.22-0ubuntu0~gpu16.04.2                   amd64        NVIDIA OpenCL ICD
ii  nvidia-opencl-icd-387                       387.34-0ubuntu0~gpu16.04.2                   amd64        NVIDIA OpenCL ICD
ii  nvidia-prime                                0.8.2                                        amd64        Tools to enable NVIDIA's Prime
ii  nvidia-settings                             387.22-0ubuntu0~gpu16.04.1                   amd64        Tool for configuring the NVIDIA graphics driver

**Comment by creatornader** [comment]
And when I run "dpkg -l | grep nvidia" I find that I have all the latest drivers installed (nvidia-387 as well as nvidia-modprobe).

As far as your advice... "I would imagine disabling the i915 and ipts modules, and associating the display with your nvidia driver will make it work." How would I go about doing this in practice? Sorry, kind of a beginner when it comes to custom Linux kernels.

**Comment by jakeday** [comment]
@creatornader You could either blacklist the modules for i915 and ipts, or set the display and driver setup in your xorg.conf files.

**Comment by jakeday** [comment]
Closing as the device is detected and loading correctly, but you'll need to set your display to use it and disable the i915 and ipts modules.

**Comment by didibaba** [comment]
I tried the "nvidia-smi" command which is used to indicate the status of the GPU usage, but it didn't work. So, I still think that there is problem with the nvidia driver.

**Comment by bareya** [comment]
Hello @didibaba I use SB2 13'' for VFX 3D programming and I can tell you it's working.

If anyone of you, ends up here as I did, here is what I can recommend:
 * Use latest drivers from [launchpad](https://launchpad.net/~graphics-drivers/+archive/ubuntu/ppa)
 * Check if `nvidia` device is recognized. There are multiple ways to do that: `lshw -C display`, `hwinfo --gfxcard`.
 * Check with `lsmod | grep nvidia` if `nvidia` module has been installed properly and if appears in the modules list. I tried to install `nvidia` drivers on `bionic beaver` and for some reason `nvidia` kernel module failed to be installed. `nvidia.ko` kernel object file was nowhere in the system. I flipped `bionic beaver` multiple times to make it working.
 * Check what's going on in `/var/log/Xorg.0.log`, sometimes you get information that your distribution fails to load `nvidia` module, then you know where to look for. 
 * @jakeday it's not necessary to blacklist Intel's driver nor manual edit xorg configuration. Instead, I use nvidia's wrapper `sudo prime-select nvidia` and `sudo prime-select intel`. It requires a reboot, but I am not a princess. I tried to make it working with `bumblebee`, but I'd have to work on it a bit longer.
